coord/coordinator.go

```diff
@@ -15,30 +15,44 @@ import (
 	"github.com/plprobelab/go-kademlia/network/address"
 	"github.com/plprobelab/go-kademlia/network/endpoint"
 	"github.com/plprobelab/go-kademlia/query"
+	"github.com/plprobelab/go-kademlia/routing"
 	"github.com/plprobelab/go-kademlia/util"
 )
 
+// A StateMachine progresses through a set of states in response to transition events.
+type StateMachine[S any, E event.Action] interface {
+	// Enqueue enqueues an event to be processed by the state machine.
+	Enqueue(context.Context, E)
+	// Advance advances the state of the state machine.
+	Advance(context.Context) S
+}
+
 // A Coordinator coordinates the state machines that comprise a Kademlia DHT
-// Currently this is only queries but will expand to include other state machines such as routing table refresh,
-// and reproviding.
+// Currently this is only queries and bootstrapping but will expand to include other state machines such as
+// routing table refresh, and reproviding.
 type Coordinator[K kad.Key[K], A kad.Address[A]] struct {
 	// self is the node id of the system the coordinator is running on
 	self kad.NodeID[K]
 
 	// cfg is a copy of the optional configuration supplied to the coordinator
 	cfg Config
 
-	qp *query.Pool[K, A]
+	// pool is the query pool state machine, responsible for running user-submitted queries
+	pool StateMachine[query.PoolState, query.PoolEvent]
+
+	// bootstrap is the bootstrap state machine, responsible for bootstrapping the routing table
+	bootstrap StateMachine[routing.BootstrapState, routing.BootstrapEvent]
```

Comment:
We could define the StateMachine type in `routing`, e.g.:

```go
type BootstrapStateMachine coord.StateMachine[BootstrapState, BootstrapEvent]
```

(cyclic dependency as the packages are structured now though)

Reply:
I prefer to define interfaces where they are used. In this case the coordinator is using the StateMachine interface.